worker/codalabworker/aws_batch.py

```diff
@@ -0,0 +1,539 @@
+                             dependencies=self._dependencies)
+        self.logger.info("Job %s transitioning %s -> %s", self.uuid, self.name, new_state.name)
+        return new_state
+
+    def update_metadata(self, **kwargs):
+        self.logger.debug("Updating metadata: %s", kwargs)
+        # Update the bundle locally
+        self._bundle['metadata'].update(kwargs)
+        # Update the bundle on the bundle service
+        self._bundle_service.update_bundle_metadata(self._worker.id, self._bundle['uuid'], kwargs)
+
+
+class Initial(AwsBatchRunState):
+    def update(self):
+        # If this job has previously made some progress, then pickup where we left off
+        if self.metadata.get('batch_job_id'):
+            transition = self.transition(Running)
+        elif self.metadata.get('batch_job_definition'):
+            transition = self.transition(Submit)
+        else:
+            transition = self.transition(Setup)
+
+        return transition
+
+
+class Setup(AwsBatchRunState):
+    def update(self):
+        # Create job definition
+        job_definition = self.create_job_definition(self._dependencies)
+
+        self.update_metadata(batch_job_definition=job_definition)
+        return self.transition(Submit)
+
+    @property
+    def docker_image(self):
+        return self._bundle['metadata']['docker_image']
+
+    @property
+    def docker_working_directory(self):
+        return '/' + self.uuid
+
+    @property
+    def docker_command(self):
+        bash_commands = [
+            'cd %s' % self.docker_working_directory,
+            '(%s) >stdout 2>stderr' % self._bundle['command'],
+            ]
+
+        return ['bash', '-c', '; '.join(bash_commands)]
+
+    def create_job_definition(self, dependencies):
+        """
+        Create the Batch job definition.
+        Each run has its own job definition which it cleans up when the run is complete.
+        """
+        bundle = self._bundle
+        # The docker image is always specified
+        image = self.resources['docker_image']
+        # Default to 100 MB so we have some breathing room.
+        memory_bytes = self.resources.get('request_memory') or 100*BYTES_PER_MEGABYTE
+        cpus = max(self.resources.get('cpus', 0), 1)
+
+        # Note: The bundle path MUST be on a shared mount between the worker machine and the compute environment nodes.
+        volumes_and_mounts = [self.volume_and_mount(
+            host_path=self._bundle_path,
+            container_path=self.docker_working_directory,
+            name=self.uuid,
+            read_only=False
+        )]
+        for host_path, docker_path, uuid in dependencies:
+            volumes_and_mounts.append(self.volume_and_mount(
+                host_path=host_path,
+                container_path=docker_path,
+                name='dependency_'+uuid,
+                read_only=True
+            ))
+
+        job_definition = {
+            'jobDefinitionName': bundle['uuid'],
```

Comment:
Names aren't unique, so I don't think collisions are a problem. I personally find the bundle uuid easiest because then you can copy-paste into codalab interfaces w/o any changes.

Reply:
I kinda feel like being able to tell they are from codalab would be good in case we end up making Batch jobs for other things, too. (So they at least all sort together in the UI.) Tags would be better but that seems to not be a thing.

Reply:
Jobs are always associated with the queue, and imo this is the job of the queue to list this association. e.g. name a queue `codalab-general` and only submit Codalab things to it. Given that queues have a many-to-one relationship with compute environments, it would be just plain odd to have a queue you submit both codalab jobs to and other jobs.

Reply:
Also, if we have other processes which end up submitting jobs which visually looks like Codalab uuids, I will be very very sad.

Reply:
Jobs are associated with a queue, but this is job definitions, which are global to our AWS account, so I think namespacing of some kind is important.

Reply:
(Also are you sure they're not unique by name?)

Reply:
This relates to my updated response below. The more I thought about it, the more I am convinced that the job ids should go on the bundle and should be treated essentially as having 1-per-run-bundle.

Reply:
When you register a job definition for a name which already exists, it creates a new revision and returns your the jobDefinitionArn for that revision. This will be something like `...:job-definition/0x25b12f80c09548f1a72358d75ea89ddd:2` for Revision 2 and bundle `0x25b12f80c09548f1a72358d75ea89ddd`

Reply:
p.s. I was also not reading the location in the code properly. I'm okay with renaming the job definitions to have a prefix. I misinterpreted this comment to be about the job itself.

If all you care about is the job definition, and you feel strongly, then I am happy to change. (just give me a 👍 )

Reply:
ok, yeah, that makes more sense. I think jobs can be whatever but I do think we should namespace the job definition, thanks.

Reply:
After more reflection, I decided my objections were silly and I should just listen to your wisdom in the first place. I've changed the naming convention for both job definitions and runs to `codalab-worker-$VERSION-$UUID`